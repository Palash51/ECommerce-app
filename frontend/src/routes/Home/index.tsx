import React, { useEffect, FC } from 'react';
import { RootState } from 'typesafe-actions';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { listProducts } from '../../reducers/products/actions';
import { IProduct } from '../../interfaces/entities';

const ProductWrapper = styled.ul`
    display: flex;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
`;

const Home: FC = () => {
    const productList = useSelector((state: RootState) => state.productList);
    const { products, loading, error } = productList;
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(listProducts());
        // return () => { 
        // }
    }, [])

    return (
        loading ? <div>loading...</div> :
            error ? <div>{error}</div> :
                <ProductWrapper>
                    {
                        products.map((product: IProduct) =>
                            <li style={{ listStyleType: "none" }} key={product._id}>
                                <div className="product">
                                    <Link to={`/product/${product._id}`} href="!#">
                                        <img className="product-image" src={product.image} alt="products" />
                                    </Link>
                                    <div className="product-name">
                                        <Link to={`/product/${product._id}`}>{product.name}</Link>
                                    </div>
                                    <div className="product-brand">{product.brand}</div>
                                    <div className="product-price">{product.price}</div>
                                    <div className="product-rating">{product.rating} stars ({product.numReviews} Reviews)</div>
                                </div>
                            </li>
                        )}
                </ProductWrapper>
    );
}

export default Home;