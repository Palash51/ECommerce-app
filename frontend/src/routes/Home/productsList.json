{
    "products":[
            {
                "brand": "Nike",
                "category": "Shirts",
                "countInStock": 6,
                "image": "/images/d1.jpg",
                "name": "Slim Shirts",
                "numReviews": 10,
                "price": 60,
                "rating": 4.5,
                "_id": "1"
            },
            {
                "brand": "Levis",
                "category": "Shirts",
                "countInStock": 8,
                "image": "/images/d1.jpg",
                "name": "Slim Shirts",
                "numReviews": 10,
                "price": 100,
                "rating": 3.5,
                "_id": "2"
            }
    ]
}
