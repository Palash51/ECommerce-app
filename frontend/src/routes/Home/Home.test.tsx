import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store'
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk'
import Home from './index';
import productsList from './productsList.json'


const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)


describe('Home component', () => {

    it('Should match snapshot', () => {
        const store = mockStore({   
            productList: productsList
        })
        const wrapper = mount(
            <Provider store={store}>
              <BrowserRouter>
              <Home />
            </BrowserRouter>
            </Provider>,
        )
        expect(wrapper).toMatchSnapshot();
    })

    it('Test products listing on home page', () => {
        const store = mockStore({   
            productList: productsList
        })
        const wrapper = mount(
            <Provider store={store}>
              <BrowserRouter>
              <Home />
            </BrowserRouter>
            </Provider>,
        )
        expect(wrapper.find('[data-testid="products-list"]')).toHaveLength(2)
    })    
})

