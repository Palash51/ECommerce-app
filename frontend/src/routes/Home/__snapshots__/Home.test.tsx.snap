// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Home>
        <styled.ul>
          <ul
            className="sc-bdnylx jVRekd"
          >
            <li
              data-testid="products-list"
              key="1"
              style={
                Object {
                  "listStyleType": "none",
                }
              }
            >
              <div
                className="product"
              >
                <Link
                  href="!#"
                  to="/product/1"
                >
                  <LinkAnchor
                    href="/product/1"
                    navigate={[Function]}
                  >
                    <a
                      href="/product/1"
                      onClick={[Function]}
                    >
                      <img
                        alt="products"
                        className="product-image"
                        src="/images/d1.jpg"
                      />
                    </a>
                  </LinkAnchor>
                </Link>
                <div
                  className="product-name"
                >
                  <Link
                    to="/product/1"
                  >
                    <LinkAnchor
                      href="/product/1"
                      navigate={[Function]}
                    >
                      <a
                        href="/product/1"
                        onClick={[Function]}
                      >
                        Slim Shirts
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
                <div
                  className="product-brand"
                >
                  Nike
                </div>
                <div
                  className="product-price"
                >
                  60
                </div>
                <div
                  className="product-rating"
                >
                  4.5
                   stars (
                  10
                   Reviews)
                </div>
              </div>
            </li>
            <li
              data-testid="products-list"
              key="2"
              style={
                Object {
                  "listStyleType": "none",
                }
              }
            >
              <div
                className="product"
              >
                <Link
                  href="!#"
                  to="/product/2"
                >
                  <LinkAnchor
                    href="/product/2"
                    navigate={[Function]}
                  >
                    <a
                      href="/product/2"
                      onClick={[Function]}
                    >
                      <img
                        alt="products"
                        className="product-image"
                        src="/images/d1.jpg"
                      />
                    </a>
                  </LinkAnchor>
                </Link>
                <div
                  className="product-name"
                >
                  <Link
                    to="/product/2"
                  >
                    <LinkAnchor
                      href="/product/2"
                      navigate={[Function]}
                    >
                      <a
                        href="/product/2"
                        onClick={[Function]}
                      >
                        Slim Shirts
                      </a>
                    </LinkAnchor>
                  </Link>
                </div>
                <div
                  className="product-brand"
                >
                  Levis
                </div>
                <div
                  className="product-price"
                >
                  100
                </div>
                <div
                  className="product-rating"
                >
                  3.5
                   stars (
                  10
                   Reviews)
                </div>
              </div>
            </li>
          </ul>
        </styled.ul>
      </Home>
    </Router>
  </BrowserRouter>
</Provider>
`;
